registry-library/library/library.go

```diff
@@ -0,0 +1,190 @@
+package library
+
+import (
+	"archive/tar"
+	"compress/gzip"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"os"
+	"path"
+	"text/tabwriter"
+	"time"
+
+	"github.com/containerd/containerd/remotes/docker"
+	"github.com/deislabs/oras/pkg/content"
+	"github.com/deislabs/oras/pkg/oras"
+	indexSchema "github.com/devfile/registry-support/index/generator/schema"
+)
+
+const (
+	// Devfile media types
+	DevfileConfigMediaType  = "application/vnd.devfileio.devfile.config.v2+json"
+	DevfileMediaType        = "application/vnd.devfileio.devfile.layer.v1"
+	DevfileVSXMediaType     = "application/vnd.devfileio.vsx.layer.v1.tar"
+	DevfileSVGLogoMediaType = "image/svg+xml"
+	DevfilePNGLogoMediaType = "image/png"
+	DevfileArchiveMediaType = "application/x-tar"
+
+	scheme                = "http"
+	httpRequestTimeout    = 30 * time.Second // httpRequestTimeout configures timeout of all HTTP requests
+	responseHeaderTimeout = 30 * time.Second // responseHeaderTimeout is the timeout to retrieve the server's response headers
+)
+
+var (
+	stacksPath = os.Getenv("DEVFILE_STACKS")
+	indexPath  = os.Getenv("DEVFILE_INDEX")
+)
+
+// GetRegistryIndex gets the index of the specified registry
+func GetRegistryIndex(registry string) ([]indexSchema.Schema, error) {
+	// Call index server REST API to get the index
+	url := scheme + "://" + path.Join(registry, "index")
+	req, err := http.NewRequest("GET", url, nil)
+	if err != nil {
+		return nil, err
+	}
+	httpClient := &http.Client{
+		Transport: &http.Transport{
+			ResponseHeaderTimeout: responseHeaderTimeout,
+		},
+		Timeout: httpRequestTimeout,
+	}
+	resp, err := httpClient.Do(req)
+	if err != nil {
+		return nil, err
+	}
+	bytes, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, err
+	}
+	var registryIndex []indexSchema.Schema
+	err = json.Unmarshal(bytes, &registryIndex)
+	if err != nil {
+		return nil, err
+	}
+	return registryIndex, nil
+}
+
+// ListRegistryStacks lists the stacks of the registry
+func ListRegistryStacks(registry string) error {
+	// Get the registry index
+	registryIndex, err := GetRegistryIndex(registry)
+	if err != nil {
+		return err
+	}
+
+	w := tabwriter.NewWriter(os.Stdout, 5, 2, 3, ' ', tabwriter.TabIndent)
+	fmt.Fprintln(w, "Name", "\t", "Description", "\t")
+	for _, stackIndex := range registryIndex {
+		fmt.Fprintln(w, stackIndex.Name, "\t", stackIndex.Description, "\t")
+	}
+	w.Flush()
+	return nil
+}
+
+// PullStackFromRegistry pulls stack from registry with allowed media types
+func PullStackFromRegistry(registry string, stack string, allowedMediaTypes []string) error {
```

Comment:
To be honest, I'm not crazy about the `PullStackFromRegistry` function requiring users to specify the media types. Most consumers will only ever need to pull all of the files in the stack

 It should always use `devfileAllMediaTypes` and maybe we should have a separate function that allows users/devs to specify the media types they want pulled

Reply:
Given golang doesn't support optional argument for function, instead of creating separate function to support the customized allowed media types, we may use the struct to wrap all the function arguments then pass the struct to the function, by default the `allowedMediaTypes` is `devfileAllMediaTypes` if the `allowedMediaTypes` is not specified in the struct.

Reply:
That still seems confusing to people using the registry library though

Why not just do something like this, and then code isn't really duplicated?
```
func PullStackByMediaTypesFromRegistry(registry string, allowedMediaTypes []string) error)
func PullStackFromRegistry(registry string) {
  return PullStackByMediaTypesFromRegistry(registry, devfileAllMediaTypes)
}
```

Reply:
That's fine for me, I can do the following function signatures, also add the missing `stack` argument

```
// Public function so that consumer can directly call it directly by case
func PullStackByMediaTypesFromRegistry(registry string, stack string, allowedMediaTypes []string) error {
    // As above...
} 

func PullStackFromRegistry(registry string, stack string) error {
  return PullStackByMediaTypesFromRegistry(registry, stack, devfileAllMediaTypes)
}
```